al/auxeffectslot.cpp

```diff
@@ -1042,3 +1046,840 @@ EffectSlotSubList::~EffectSlotSubList()
+
+        case 1:
+            eax_set_default_slot_1_defaults();
+            break;
+
+        case 2:
+        case 3:
+            break;
+
+        default:
+            fail("FX slot index out of range.");
+    }
+}
+
+void ALeffectslot::eax_get_fx_slot_all(
+    const EaxEaxCall& eax_call) const
+{
+    switch (eax_call.get_version())
+    {
+        case 4:
+            eax_call.set_value<EaxFxSlotException, EAX40FXSLOTPROPERTIES>(eax_eax_fx_slot_);
+            break;
+
+        case 5:
+            eax_call.set_value<EaxFxSlotException, EAX50FXSLOTPROPERTIES>(eax_eax_fx_slot_);
+            break;
+
+        default:
+            fail("Unsupported EAX version.");
+    }
+}
+
+void ALeffectslot::eax_get_fx_slot(
+    const EaxEaxCall& eax_call) const
+{
+    switch (eax_call.get_property_id())
+    {
+        case EAXFXSLOT_ALLPARAMETERS:
+            eax_get_fx_slot_all(eax_call);
+            break;
+
+        case EAXFXSLOT_LOADEFFECT:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.guidLoadEffect);
+            break;
+
+        case EAXFXSLOT_VOLUME:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.lVolume);
+            break;
+
+        case EAXFXSLOT_LOCK:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.lLock);
+            break;
+
+        case EAXFXSLOT_FLAGS:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.ulFlags);
+            break;
+
+        case EAXFXSLOT_OCCLUSION:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.lOcclusion);
+            break;
+
+        case EAXFXSLOT_OCCLUSIONLFRATIO:
+            eax_call.set_value<EaxFxSlotException>(eax_eax_fx_slot_.flOcclusionLFRatio);
+            break;
+
+        default:
+            fail("Unsupported FX slot property id.");
+    }
+}
+
+// [[nodiscard]]
+bool ALeffectslot::eax_get(
+    const EaxEaxCall& eax_call)
+{
+    switch (eax_call.get_property_set_id())
+    {
+        case EaxEaxCallPropertySetId::fx_slot:
+            eax_get_fx_slot(eax_call);
+            break;
+
+        case EaxEaxCallPropertySetId::fx_slot_effect:
+            eax_dispatch_effect(eax_call);
+            break;
+
+        default:
+            fail("Unsupported property id.");
+    }
+
+    return false;
+}
+
+void ALeffectslot::eax_set_fx_slot_effect(
+    EaxAlEffectUPtr& effect)
+{
+    eax_al_effect_ = effect.get();
+
+    eax_set_effect_slot_effect(*eax_al_effect_);
+}
+
+void ALeffectslot::eax_set_fx_slot_effect(
+    ALenum al_effect_type,
+    EaxAlEffectUPtr& effect)
+{
+    if (!effect)
+    {
+        effect = eax_create_al_effect(*eax_al_context_, al_effect_type);
+        effect->eax_initialize();
+    }
+
+    eax_set_fx_slot_effect(effect);
+}
+
+void ALeffectslot::eax_set_fx_slot_effect()
+{
+    if (false)
+    {
+    }
+    else if (eax_eax_fx_slot_.guidLoadEffect == EAX_NULL_GUID)
+    {
+        eax_set_fx_slot_effect(AL_EFFECT_NULL, eax_null_effect_);
```

Comment:
Is it intended that when an fxslot changes its effect type (e.g. chorus to reverb, or reverb to null), the parameter values for the previous effect type are retained for when it's changed back?

Reply:
It's not. I'll fix that.

Reply:
It should be a fairly easy fix. Rather than a bunch of `EaxAlEffectUPtr eax_<type>_effect_{};` with one `ALeffect* eax_al_effect_{};` that points to the "current" one, it just needs one `eax_al_effect_`, and changing the effect type sets all properties for that type to defaults (which then get applied on the effect object as normal).